Framework/NexusGeometry/test/NexusGeometrySaveTest.h

```diff
@@ -0,0 +1,1539 @@
+      }
+    }
+  }
+
+  void
+  test_when_nx_source_group_has_nx_transformations_transformation_type_is_specified() {
+
+    // create RAII file resource for testing
+    ScopedFileHandle fileResource(
+        "check_nxsource_group_has_transformation_type_test_file.hdf5");
+    std::string destinationFile = fileResource.fullPath();
+
+    // call saveInstrument
+    NexusGeometrySave::saveInstrument(m_instrument, destinationFile,
+                                      DEFAULT_ROOT_PATH);
+
+    // get Instrument cache from default unit test instrument
+    auto &compInfo = (*m_instrument.first);
+
+    // get component names to access path to H5 group
+    auto instrName = compInfo.name(compInfo.root());
+    auto sourceName = compInfo.name(compInfo.source());
+
+    // instance of test utility to check saved file
+    HDF5FileTestUtility tester(destinationFile);
+
+    // full path to group to be opened in test utility
+    fullH5Path path = {DEFAULT_ROOT_PATH, instrName, sourceName,
+                       TRANSFORMATIONS};
+
+    // assertations
+    bool hasNXTransformation =
+        tester.hasAttributeInGroup(NX_CLASS, NX_TRANSFORMATIONS, path);
+    bool hasTranslation = tester.hasDataset(LOCATION, path);
+    bool hasRotation = tester.hasDataset(ORIENTATION, path);
+    bool hasEither = (hasRotation || hasTranslation);
+    // TODO: having such a group may be optional.
+    TS_ASSERT(hasNXTransformation);
+    TS_ASSERT(hasEither);
+  }
+
+  void
+  test_xyz_pixel_offset_in_file_is_relative_position_from_bank_without_rotation() {
+
+    // create RAII file resource for testing
+    ScopedFileHandle fileResource("check_pixel_offset_format_test_file.hdf5");
+    std::string destinationFile = fileResource.fullPath();
+
+    // prepare geometry for instrument
+    const Quat relativeBankRotation(45.0, V3D(0.0, 1.0, 0.0));
+    const Quat relativeDetRotation(45.0, V3D(0.0, 1.0, 0.0));
+    const V3D absBankposition(0, 0, 10);
+    const V3D detPosition(2.0, -2.0, 0.0);
+
+    // create test instrument and get cache
+    auto instrument =
+        ComponentCreationHelper::createSimpleInstrumentWithRotation(
+            Mantid::Kernel::V3D(0, 0, -10), Mantid::Kernel::V3D(0, 0, 0),
+            absBankposition,      // bank position
+            relativeBankRotation, // bank rotation
+            relativeDetRotation,  // detector rotation
+            detPosition);         // detector position
+    auto instr = Mantid::Geometry::InstrumentVisitor::makeWrappers(*instrument);
+    auto &compInfo = (*instr.first);
+
+    // saveinstrument
+    NexusGeometrySave::saveInstrument(instr, destinationFile,
+                                      DEFAULT_ROOT_PATH);
+
+    // get instrument name to access H5 group in test utility
+    auto instrName = compInfo.name(compInfo.root());
+
+    // instance of test utility to check saved file
+    HDF5FileTestUtility tester(destinationFile);
+
+    for (size_t idx = compInfo.root() - 1; idx > 0; --idx) {
```

Comment:
Explain this loop for readers.

- Why ignore root?
- Why reverse iterate?
- Why go to idx = 1 (actually contains a bug and a missing optimisation)

Reply:
refactored